File: src/main/java/com/leet/medium/PowXN.java
package com.leet.medium;

import java.util.stream.Collectors;
import java.util.stream.Stream;

public class PowXN {
  public double myPow(double x, int n) {
    if (x == 1) {
      return 1;
    }

    double ans = 1, base = n < 0 ? 1d/x : x;
    int pow = n < 0 ? -n : n;
    while (pow != 0) {
      if ((pow & 1) == 1) {
        ans *= base;
      }
      if (base >= ~(Integer.MIN_VALUE >>> 1)) {
        return 0;
      }
      pow >>= 1;
      base = base * base;
    }

    return ans;
  }
}

File: src/main/java/com/leet/medium/PowXN.go
package main

import (
	"fmt"
	"math"
)

func myPow(x float64, n int) float64 {
	if x == float64(1) {
		return 1
	}
	base := x
	if n < 0 {
		base = 1 / x
		n *= -1
	}
	ans, pow := float64(1), n
	for pow != 0 {
		if pow&1 == 1 {
			ans = ans * base
		}
		pow >>= 1
		base = base * base
	}
	return ans
}

func main() {
	fmt.Println(myPow(2, 31), math.Pow(2, 31))
}
